### PR TITLE
DMP Queue pallet

### PR DESCRIPTION
**Pull request by gavofyork**
Part of https://github.com/paritytech/polkadot/issues/2841.

A bit similar to the XCMP Queue pallet, but for DMP messages.

- [x] Tests
  - [x] Service-queued (weight limit greater than queued message weight)
  - [x] Service-inline (nothing queued, weight limit greater than incoming message weight)
  - [x] Enqueue (nothing queued, weight limit less than incoming message weight)
  - [x] Service-inline-then-Enqueue (nothing queued, multiple incoming messages, weight limit greater than first incoming message weight, less than total incoming message weight)
  - [x] Service-queued-then-service-inline (incoming message(s) and queued messages(s); weight-limit greater than total of all messages' weight)
  - [x] Service-queued-partially-then-enqueue (queued messages and incoming messages; weight-limit less than total of queued messages weight and incoming message weight)
  - [x] Service-queued-completely-then-enqueue (queued messages and incoming messages; weight-limit greater than total of queued messages weight but less than total of queued messages weight plus the first incoming message weight)
  - [x] Service-queued-then-service-inline-then-enqueue (queued messages and incoming messages; weight limit greater than queued message weights plus one incoming message's weight but less than sum of all message weights)
  - [x] Overweight should queue (overweight messages should be enqueued in the above situations without affecting results)
  - [x] Overweight can be dispatched be correct origin
  - [x] OnIdle should service queue

### CODE REVIEW

File: pallets/dmp-queue/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,782 @@
...
+			weight_limit: Weight,
+		) -> DispatchResultWithPostInfo {
+			T::ExecuteOverweightOrigin::ensure_origin(origin)?;
+
+			let (sent_at, data) = Overweight::<T>::get(index).ok_or(Error::<T>::Unknown)?;
+			let used = Self::try_service_message(weight_limit, sent_at, &data[..])
+				.map_err(|_| Error::<T>::OverLimit)?;
+			Overweight::<T>::remove(index);
+			Self::deposit_event(Event::OverweightServiced(index, used));
+			Ok(PostDispatchInfo { actual_weight: Some(1_000_000 + used), pays_fee: Pays::Yes })
```
Comment (apopiak) [inline pallets/dmp-queue/src/lib.rs:162]:
```suggestion
			Ok(Some(1_000_000 + used).into())
```
should work I think?

File: pallets/dmp-queue/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+		/// Downward message is invalid XCM.
+		/// \[ id \]
+		InvalidFormat(MessageId),
+		/// Downward message is unsupported version of XCM.
+		/// \[ id \]
+		UnsupportedVersion(MessageId),
+		/// Downward message executed with the given outcome.
+		/// \[ id, outcome \]
+		ExecutedDownward(MessageId, Outcome),
+		/// \[ id, remaining, required \]
```
Comment (apopiak) [inline pallets/dmp-queue/src/lib.rs:179]:
```suggestion
		/// The weight limit for handling downward messages was reached.
		/// \[ id, remaining, required \]
```

File: pallets/dmp-queue/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+		WeightExhausted(MessageId, Weight, Weight),
+		/// \[ id, index, required \]
```
Comment (apopiak) [inline pallets/dmp-queue/src/lib.rs:181]:
```suggestion
		/// Downward message is overweight and was placed in the overweight queue.
		/// \[ id, index, required \]
```

File: pallets/dmp-queue/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+		OverweightEnqueued(MessageId, OverweightIndex, Weight),
+		/// \[ index, used \]
```
Comment (apopiak) [inline pallets/dmp-queue/src/lib.rs:183]:
```suggestion
		/// Downward message from the overweight queue was executed.
		/// \[ index, used \]
```

File: pallets/dmp-queue/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,782 @@
...
+		}
+
+		/// Exactly equivalent to `service_queue` but expects a mutable `page_index` to be passed
+		/// in and any changes stored.
+		fn do_service_queue(limit: Weight, page_index: &mut PageIndexData) -> Weight {
+			let mut used = 0;
+			while page_index.begin_used < page_index.end_used {
+				let page = Pages::<T>::take(page_index.begin_used);
+				for (i, &(sent_at, ref data)) in page.iter().enumerate() {
+					match Self::try_service_message(limit - used, sent_at, &data[..]) {
```
Comment (apopiak) [inline pallets/dmp-queue/src/lib.rs:202]:
```suggestion
					match Self::try_service_message(limit.saturating_sub(used), sent_at, &data[..]) {
```

File: pallets/dmp-queue/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,782 @@
...
+		/// - `index`: The index of the overweight message to service.
+		/// - `weight_limit`: The amount of weight that message execution may take.
+		///
+		/// Errors:
+		/// - `Unknown`: Message of `index` is unknown.
+		/// - `OverLimit`: Message execution may used greater than `weight_limit`.
+		///
+		/// Events:
+		/// - `OverweightServiced`: On success.
+		#[pallet::weight(1_000_000 + weight_limit)]
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply pallets/dmp-queue/src/lib.rs:149]: yup - i made an issue for it.

**Comment by gavofyork** [comment]
@apopiak final approval?

**Review by apopiak — APPROVED** [review]

File: pallets/dmp-queue/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+	#[pallet::storage]
+	pub(super) type PageIndex<T> = StorageValue<_, PageIndexData, ValueQuery>;
+
+	/// The queue pages.
+	#[pallet::storage]
+	pub(super) type Pages<T> = StorageMap<
+		_,
+		Blake2_128Concat,
+		PageCounter,
+		Vec<(RelayBlockNumber, Vec<u8>)>,
```
Comment (shawntabrizi) [inline pallets/dmp-queue/src/lib.rs:106]: is it possible to already use `BoundedVec` here?

Would prevent having to backport those changes

File: pallets/dmp-queue/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+		ValueQuery,
+	>;
+
+	/// The overweight messages.
+	#[pallet::storage]
+	pub(super) type Overweight<T> = StorageMap<
+		_,
+		Blake2_128Concat,
+		OverweightIndex,
+		(RelayBlockNumber, Vec<u8>),
```
Comment (shawntabrizi) [inline pallets/dmp-queue/src/lib.rs:116]:
```suggestion
		(RelayBlockNumber, Vec<u8>),
```

here too

File: pallets/dmp-queue/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+		/// - `index`: The index of the overweight message to service.
+		/// - `weight_limit`: The amount of weight that message execution may take.
+		///
+		/// Errors:
+		/// - `Unknown`: Message of `index` is unknown.
+		/// - `OverLimit`: Message execution may use greater than `weight_limit`.
+		///
+		/// Events:
+		/// - `OverweightServiced`: On success.
+		#[pallet::weight(1_000_000 + weight_limit)]
```
Comment (shawntabrizi) [inline pallets/dmp-queue/src/lib.rs:150]:
```suggestion
		#[pallet::weight(1_000_000.saturating_add(weight_limit))]
```

File: pallets/dmp-queue/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,786 @@
...
+			weight_limit: Weight,
+		) -> DispatchResultWithPostInfo {
+			T::ExecuteOverweightOrigin::ensure_origin(origin)?;
+
+			let (sent_at, data) = Overweight::<T>::get(index).ok_or(Error::<T>::Unknown)?;
+			let used = Self::try_service_message(weight_limit, sent_at, &data[..])
+				.map_err(|_| Error::<T>::OverLimit)?;
+			Overweight::<T>::remove(index);
+			Self::deposit_event(Event::OverweightServiced(index, used));
+			Ok(Some(1_000_000 + used).into())
```
Comment (shawntabrizi) [inline pallets/dmp-queue/src/lib.rs:163]:
```suggestion
			Ok(Some(1_000_000.saturating_add(used)).into())
```

File: primitives/core/src/lib.rs [hunk 11]
```diff
@@ -118,12 +118,23 @@ pub mod well_known_keys {
...
+	) -> Weight;
 }
-impl DownwardMessageHandler for () {
-	fn handle_downward_message(_msg: InboundDownwardMessage) -> Weight { 0 }
+impl DmpMessageHandler for () {
+	fn handle_dmp_messages(
+		iter: impl Iterator<Item=(RelayBlockNumber, Vec<u8>)>,
+		_max_weight: Weight,
+	) -> Weight {
+		for _ in iter {}
```
Comment (shawntabrizi) [inline primitives/core/src/lib.rs:135]: maybe dumb question, but this gets optimized away?

**Review by shawntabrizi — APPROVED** [review]
high level review looks good

File: primitives/core/src/lib.rs (hunk 11 above)
Comment (gavofyork) [reply primitives/core/src/lib.rs:135]: I don't think so as `next()` can have side effects.

File: pallets/dmp-queue/src/lib.rs (hunk 10 above)
Comment (gavofyork) [reply pallets/dmp-queue/src/lib.rs:163]: `used` must be sensible, so there's no need for saturating here.

File: pallets/dmp-queue/src/lib.rs (hunk 7 above)
Comment (gavofyork) [reply pallets/dmp-queue/src/lib.rs:106]: will take a while yet :)

**Pull request merged**